src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3773,8 +3772,15 @@ void os::large_page_init() {
 
   // Populate _page_sizes with large page sizes less than or equal to
   // _large_page_size.
+  bool large_page_found = false;
   for (size_t page_size = _large_page_size; page_size != 0;
          page_size = all_large_pages.next_smaller(page_size)) {
+    if (!large_page_found &&
+      page_size != (size_t)os::vm_page_size() &&
+      os::Linux::setup_large_page(page_size)) {
+      _large_page_size = page_size;
+      large_page_found = true;
```

Comment:
I don't think this is at the right level: here we don't yet know which kind of large page to use, which is decided in `os::Linux::setup_large_page_type` (the original name).

Instead, the fix should be inside `hugetlbfs_sanity_check`, sth like:

```cpp
for (local_page_size = page_size; page_size_ != (size_t)os::vm_page_size(); local_page_size = _page_sizes.next_smaller(local_page_size) {
  if (try_commit_using_large_page(local_page_size)) {
    _large_page_size = local_page_size;
    return true;
  }
}

return false;
```